### DISCUSSION TITLE
Enabling Player Hitbox with emulator?

### DISCUSSION THREAD

**Discussion by Rosenstein** [Drek282/PsychoStats#8] [Ideas]
Have you tought about using [Flash Player emulator](https://ruffle.rs/#usage) for player hitboxes as "temporary" fix until somebody ports .swf file to HTML5? It's still WIP, but easy to use and works in every browser/device.

All you need is 1 line of code under `<div id="s-hitbox" class="ps-table-inner">` in ..\themes\\*\player.html after extracting files from archive in proper dir, and one .htaccess file for setting up web server to serve .wasm files correctly

![image](https://user-images.githubusercontent.com/4959837/115161135-31749b80-a09c-11eb-9767-418809e182aa.png)

**Comment by Drek282** [comment]
I wasn't aware of that.  Thanks for bringing this to my attention.  I will look into it.  Thanks for your fix for the vulnerability btw.  Much appreciated.

**Reply by Drek282** [reply]
Also, if you want to submit that code, I'll be more than happy to merge it once I figure it out.  I'll also have to add the hitbox back, I think I've removed it entirely.

**Comment by Rosenstein** [comment]
I'll see what I can do.

**Reply by Drek282** [reply]
Thank you for your submission, I'll try to look  at that today.

**Comment by Rosenstein** [comment]
While thinkering with this I've noticed console error that says: `ruffle_web.js:662 GET http://localhost/xml/config.xml 404 (Not Found)`
After that, I have found this Issue: https://github.com/d4fseeker/PsychoStats-extended/issues/20
It seems that the flash animation is looking for its config file at the wrong location ( on windows at aleast ), can you confirm?
Fixing that is easy, just copy /xml/config.xml to web root.

As for the blank place in corner, I'm looking into it but there's no ETA on that.

**Reply by Drek282** [reply]
You're getting ahead of me.  :)

I'll try to look at it today.  Wondering if copying config.xml to web root presents any conflict issues.  Can't think of any.

Considering your contribution(s) and activity over a long period of time, and considering that it is a dead certainty that you are a more advanced, knowledgeable and experienced coder than I am.  Would you like me to grant you permissions so you can work on this without my having to  approve everything?

**Reply by Rosenstein** [reply]
> I'll try to look at it today. Wondering if copying config.xml to web root presents any conflict issues. Can't think of any.

I think there are none. Only slight confusion over the placement of the file :D

> Considering your contribution(s) and activity over a long period of time, and considering that it is a dead certainty that you are a more advanced, knowledgeable and experienced coder than I am. Would you like me to grant you permissions so you can work on this without my having to approve everything?

Nope, no need for extra permissions. I'm not exactly a coder, but more of a good tinkerer, doing this stuff as a hobby.  :)
I feel like doing so will put more pressure on me, 'cause coding isn't exactly my profession.

**Reply by Drek282** [reply]
Ok, looking at that error, I'm more than a little confused.  There is no file by the name of "ruffle_web.js".  I'm going to have to test this, but not today...

**Reply by Drek282** [reply]
That file is definitely in the wrong location.  The code contained in hitbox_xml.as is looking for that file in "xml/config.xml".

**Reply by Rosenstein** [reply]
'ruffle.web.js' is irrelevant here, that's just the lib that throws an error.
We'll sort the location

**Comment by Drek282** [comment]
That config.xml file also contains an error as far as I can see.

	<bullets>12</target>
	
That line makes no sense in xml terms.  It should either be bullets /bullets or tartet /target.

**Reply by Rosenstein** [reply]
Nice catch, I didn't see that but I think it should be `<bullets>12</bullets>`.
I'll have to decompile the .swf to confirm.

**Reply by Rosenstein** [reply]
It is 'bullets', we have these defaults in swf:
```
hitbox._bullets = 10;
hitbox._title = "%s Hitbox";
```

**Reply by Drek282** [reply]
Ya, I figured that out.

**Comment by Drek282** [comment]
Those are problems with Stormtrooper's original code, and unrelated to Ruffle.

**Comment by Drek282** [comment]
There is no "xml" folder anywhere in the www/upload filesystem.  I'd say there are errors there that Stormtrooper never got around to fixing.  I'll fix the location in the script.  Putting that config.xml file in the root web folder is not the answer.

**Reply by Drek282** [reply]
My sense is that those files are not used, and are there only as place holders to suppress errors.  I will fix the references in them and fix the code, but I think it's basically irrelevant.  I don't think they actually do anything.

**Reply by Rosenstein** [reply]
Config.xml has to go somewhere, but .swf file is compiled with wrong paths.

**Comment by Drek282** [comment]
Unfortunately on testing the hitbox still doesn't show.

**Comment by Drek282** [comment]
There was a path problem.  I've almost got it working.

**Comment by Drek282** [comment]
The hitbox shows, but the only mod I run that has hit location data is TFC and there may be a problem with the mod code.  The hitbox shows but it doesn't show any data.  Do you mind testing the fixes I just committed please?

**Reply by Drek282** [reply]
The only significant change is:  <script src="{theme->parent_url}/js/ruffle/ruffle.js"></script>

You used an absolute MS path there.

**Reply by Drek282** [reply]
The problem is that the weapon list isn't being loaded.

**Reply by Rosenstein** [reply]
Can you paste browser console output when you reload player stats page? Do it on some pastebin.

**Reply by Drek282** [reply]
It's looking for xml/weapons.xml, but I know that for both config.xml and weapons.xml they shouldn't be necessary.  It should load the data from the database, and in the case of config.xml if it doesn't find the file it loads defaults.  So for some reason in this version of PS the weapons aren't being loaded from the database, and I can't figure out why.

**Comment by Rosenstein** [comment]
We can safely say that hitbox is fully enabled in your fork now.
Todo: Fix one minor issue with the hitbox loading weapon images

**Reply by Drek282** [reply]
The hitbox is supposed to load weapon images?

**Reply by Drek282** [reply]
I will look into that.

**Comment by Rosenstein** [comment]
Flash hitbox is now properly emulated as of May 1, 2023.

**Comment by Drek282** [comment]
*cheers Rosenstein*
